### PR TITLE
Particles

### PR DESCRIPTION
**Pull request by maximumpower55**

### CODE REVIEW

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/MinecraftMixin.java [hunk 1]
```diff
@@ -63,7 +72,8 @@ private void onAttack(CallbackInfoReturnable<Boolean> cir, ItemStack stack) {
 				if (result == TriState.TRUE) {
 					PortalCubedPackets.sendToServer(new DirectClickItemPacket(true, InteractionHand.MAIN_HAND));
 				}
-				cir.setReturnValue(result.toBoolean());
+				// Crowbar check or else there will be no delay between startAttack and continueAttack causing a double swing
+				cir.setReturnValue(direct instanceof CrowbarItem || result.toBoolean());
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/client/MinecraftMixin.java:76]: bad solution, why does that happen

File: src/main/java/io/github/fusionflux/portalcubed/mixin/client/ParticleEngineMixin.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	@Shadow
+	@Final
+	@Mutable
+	private static List<ParticleRenderType> RENDER_ORDER;
+
+	@Inject(method = "<clinit>", at = @At("TAIL"))
+	private static void addMultiplyRenderType(CallbackInfo ci) {
+		List<ParticleRenderType> newRenderOrder = new ArrayList<>(RENDER_ORDER);
+		newRenderOrder.add(MultiplyParticleRenderType.INSTANCE);
+		RENDER_ORDER = Collections.unmodifiableList(newRenderOrder);
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/mixin/client/ParticleEngineMixin.java:31]: don't make it immutable again

File: src/main/java/io/github/fusionflux/portalcubed/packet/clientbound/SimpleParticlePacket.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		buf.writeDouble(velocityY);
+		buf.writeDouble(velocityZ);
+	}
+
+	@Override
+	public ResourceLocation getId() {
+		return PortalCubedPackets.SIMPLE_PARTICLE;
+	}
+
+	@Override
```
Comment (TropheusJ) [inline src/main/java/io/github/fusionflux/portalcubed/packet/clientbound/SimpleParticlePacket.java:35]: missing ClientOnly

File: src/main/resources/data/portalcubed/tags/blocks/bullet_hole_metal.json [hunk 4]
```diff
@@ -43,21 +45,152 @@
     "minecraft:waxed_oxidized_cut_copper",
     "minecraft:waxed_oxidized_cut_copper_stairs",
     "minecraft:waxed_oxidized_cut_copper_slab",
+    {
+      "id": "minecraft:chiseled_copper",
+      "required": false
```
Comment (TropheusJ) [inline src/main/resources/data/portalcubed/tags/blocks/bullet_hole_metal.json:50]: pretty sure these don't need to be optional, was that tested

**Review by TropheusJ — CHANGES_REQUESTED** [review]

File: src/main/resources/data/portalcubed/tags/blocks/bullet_hole_metal.json (hunk 4 above)
Comment (Cart3r1234) [reply src/main/resources/data/portalcubed/tags/blocks/bullet_hole_metal.json]: fixed

File: src/main/java/io/github/fusionflux/portalcubed/content/fizzler/FizzleDarkParticle.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+	@NotNull
+	@Override
+	public ParticleRenderType getRenderType() {
+		return ParticleRenderType.PARTICLE_SHEET_OPAQUE;
+	}
+
+	@Override
+	public float getQuadSize(float tickDelta) {
+		return this.quadSize * (1 - Math.min((Math.max(0, (Math.min((this.age + tickDelta) / this.lifetime, 1)) - SHRINK_START_PROGRESS) / (1 - SHRINK_START_PROGRESS)), 1));
```
[the remark this answers is not in the document]
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/content/fizzler/FizzleDarkParticle.java:53]: You didn't have a problem with it when I did it for the disintegration animation :ioa:

File: src/main/java/io/github/fusionflux/portalcubed/content/misc/DecalParticle.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		if (currentBaseState.isAir()) {
+			this.remove();
+			return;
+		}
+
+		if (this.lastBaseState != null && this.lastBaseState != currentBaseState) {
+			VoxelShape baseShape = currentBaseState.getCollisionShape(this.level, this.basePos);
+			Vec3 rayStart = new Vec3(this.x, this.y, this.z);
+			Vec3 rayEnd = rayStart.subtract(Vec3.atLowerCornerOf(this.direction.getNormal()).scale(ONE_PIXEL));
+			if (Optionull
```
[the remark this answers is not in the document]
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/content/misc/DecalParticle.java:99]: Whats the problem?

File: src/main/java/io/github/fusionflux/portalcubed/content/crowbar/CrowbarItem.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+				}
+			});
+		} else if (player.isLocalPlayer()) {
+			PortalCubedPackets.sendToServer(new CrowbarSwingPacket(hit, didSwingAnim));
+		}
+	}
+
+	@Override
+	public TriState onAttack(Level level, Player player, ItemStack stack) {
+		this.onSwing(player, Minecraft.getInstance().hitResult, false);
```
[the remark this answers is not in the document]
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/content/crowbar/CrowbarItem.java:61]: Server ran fine

File: src/main/java/io/github/fusionflux/portalcubed/content/misc/DecalParticle.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/content/misc/DecalParticle.java:99]: horribly formatted extremely long multi-line if

File: src/main/java/io/github/fusionflux/portalcubed/content/crowbar/CrowbarItem.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/content/crowbar/CrowbarItem.java:62]: this is a common sided method that is supposed to get called twice. If it's not crashing, something else is wrong

[the remark this answers is not in the document]
Comment (maximumpower55) [reply src/main/java/io/github/fusionflux/portalcubed/content/crowbar/CrowbarItem.java:62]: It returns TriState.FALSE which disables the sending to the server per the javadoc

[the remark this answers is not in the document]
Comment (TropheusJ) [reply src/main/java/io/github/fusionflux/portalcubed/content/crowbar/CrowbarItem.java:62]: just add the hit result to the method arguments and packet

**Pull request merged**